### PR TITLE
Fix | Added ISQLServerBulkData to remove implementation details from ISQLServerBulkRecord

### LINKED ISSUE #1052: [FEATURE REQUEST] Remove set / add methods from ISQLServerBulkRecord
## Is your feature request related to a problem? If so, please give a short summary of the problem and how the feature would resolve it
The ISQLServerBulkRecord interface is a mixture of specification and implementation details.

## Describe the preferred solution
Remove methods which are not called as part of actual bulk copy process via SQLServerBulkCopy#writeToServer.

## Describe alternatives you've considered
Currently it's necessary to implement methods even if they just throw a UnsupportedOperationException

### PR DESCRIPTION
**Pull request by jonfreedman**
New ISQLServerBulkData interface which extracts the read-only methods from ISQLServerBulkRecord

Fixes #1052

### CODE REVIEW

**Comment by jonfreedman** [comment]
@cheenamalhotra please take a look, thanks

**Comment by jonfreedman** [comment]
Looks like the test failure is spurious - hit a deadlock

**Comment by ulvii** [comment]
Hi @jonfreedman,
Thanks for the contribution, the team will review the PR.

File: src/test/java/com/microsoft/sqlserver/jdbc/bulkCopy/BulkCopyTestUtil.java [hunk 1]
```diff
@@ -12,10 +12,7 @@
 import java.sql.ResultSetMetaData;
 import java.sql.SQLException;
 
-import com.microsoft.sqlserver.jdbc.ComparisonUtil;
-import com.microsoft.sqlserver.jdbc.ISQLServerBulkRecord;
-import com.microsoft.sqlserver.jdbc.SQLServerBulkCopy;
-import com.microsoft.sqlserver.jdbc.TestResource;
+import com.microsoft.sqlserver.jdbc.*;
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/bulkCopy/BulkCopyTestUtil.java]: Revert this change.

Comment (jonfreedman) [reply src/test/java/com/microsoft/sqlserver/jdbc/bulkCopy/BulkCopyTestUtil.java]: done

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkData.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    int getScale(int column);
+
+    /**
+     * Returns whether the column represents an identity column.
+     *
+     * @param column
+     *        Column ordinal
+     * @return True if the column is an identity column; false otherwise.
+     */
+    boolean isAutoIncrement(int column);
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkData.java]: We don't use isAutoIncrement in writeToServer, so I think we can leave this method back in ISQLServerBulkRecord.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkRecord.java [hunk 3]
```diff
@@ -15,81 +15,7 @@
...
-    /**
-     * Advances to the next data row.
-     * 
-     * @return True if rows are available; false if there are no more rows
-     * @throws SQLServerException
-     *         If there are any errors in advancing to the next row.
-     */
-    public boolean next() throws SQLServerException;
-
+public interface ISQLServerBulkRecord extends ISQLServerBulkData {
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkRecord.java]: We can remove `import java.io.Serializable;` on line 8 as we no longer need this.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkData.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ * SQLServerBulkCopy class to write the data to SQL Server tables.
+ */
+public interface ISQLServerBulkData extends Serializable {
+
+    /**
+     * Returns the ordinals for each of the columns represented in this data record.
+     *
+     * @return Set of ordinals for the columns.
+     */
+    java.util.Set<Integer> getColumnOrdinals();
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkData.java]: I know that public modifiers are redundant for interfaces, but could we add the public modifiers to the methods in this interface, just for the sake of consistency with other interfaces? Either that or please remove the public modifiers in the other interfaces as well (or you can let us handle this in the near future)

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkData.java (hunk 2 above)
Comment (jonfreedman) [reply src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkData.java]: I'll move this back

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkRecord.java (hunk 3 above)
Comment (jonfreedman) [reply src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkRecord.java]: I'll remove

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkData.java (hunk 4 above)
Comment (jonfreedman) [reply src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerBulkData.java]: I've removed from all your interfaces

**Comment by peterbae** [comment]
Looks like there's some conflicts, but other than that, the PR looks good to me.

**Comment by jonfreedman** [comment]
@peterbae I have merged from upstream

**Comment by jonfreedman** [comment]
@peterbae I've removed the cast, didn't find any others

**Comment by jonfreedman** [comment]
Can this be merged?

**Comment by rene-ye** [comment]
Hi @jonfreedman, the team is contemplating whether to merge this for the next preview or not. We're trying to decide whether we want to apply a bandaid fix or just break the public apis and change the interfaces completely. We will merge/close this PR after discussions.

**Comment by jonfreedman** [comment]
It's not really clear to me why you can't do both. You can see that this pr has been open for 3 months, based on an issue from 5 months ago. At this point all it really accomplishes is that I can delete some code that I've been running in production for almost 6 months. Are you saying that if you don't merge this then the next release will be rewritten with breaking changes, or that the 'fix' will happen at some later date?

**Comment by ulvii** [comment]
Hi @jonfreedman,
We decided to accept your suggestion, apologies for the delay. Would you mind if we make a few minor changes to your branch?

**Comment by jonfreedman** [comment]
That's great, please go ahead and change whatever you need to

**Comment by ulvii** [comment]
Hi @jonfreedman,
I created a [PR](https://github.com/jonfreedman/mssql-jdbc/pull/1) against your fork, which updates the branch from dev, adds `@deprecated` tag to ISQLServerBulkRecord, formats the files and some other minor changes. Please merge it to proceed with the review process.

**Comment by ulvii** [comment]
The PR is merged. Thanks for the contribution @jonfreedman.
